### PR TITLE
Intercom UI (FFCS)

### PR DESCRIPTION
**Pull request by TheMagnetar**
**When merged this pull request will:**
- Adds UI to Intercom (FFCS)
- Models donated by RanTa to ACRE project.

image

### CODE REVIEW

**Comment by Drofseh** [comment]
Damn that looks good

File: addons/sys_intercom/XEH_postInit.sqf [hunk 1]
```diff
@@ -40,6 +40,14 @@ if (!hasInterface) exitWith {};
     [1, false] call FUNC(switchIntercomFast)
 }, [DIK_COMMA, [false, true, true]]] call CBA_fnc_addKeybind;
 
+[
+    "ACRE2",
+    QGVAR(openGui),
+    localize LSTRING(openGui),
+    {[0] call FUNC(openGui)},
+    ""
+] call cba_fnc_addKeybind;
```
Comment (jonpas) [inline addons/sys_intercom/XEH_postInit.sqf]:
```suggestion
] call CBA_fnc_addKeybind;
```
Maybe even this for consistency:
```sqf
["ACRE2", QGVAR(openGui), localize LSTRING(openGui), {
    [0] call FUNC(openGui)
}, ""] call CBA_fnc_addKeybind;
```

File: addons/sys_intercom/fnc_closeGui.sqf [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * GUI successfully closed <BOOL>
+ *
+ * Example:
+ * [1, true] call acre_sysIntercom_fnc_openGui
+ *
+ * Public: No
+ */
+
+// The case of a player exiting a vehicle with UI opened is handled in the vehicle CBA EH
+if (vehicle acre_player isEqualTo acre_player) exitWith {};
```
Comment (jonpas) [inline addons/sys_intercom/fnc_closeGui.sqf]: Should it not return `false` though?

File: addons/sys_intercom/fnc_saveStationConfiguration.sqf [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * Public: No
+ */
+
+params ["_vehicle", "_unit", ["_varName", ""]];
+
+if (_varName isEqualTo "") then {
+    _varName = [_vehicle, _unit] call FUNC(getStationVariableName);
+};
+
+if (_varName isEqualTo "") exitWith {
```
Comment (jonpas) [inline addons/sys_intercom/fnc_saveStationConfiguration.sqf]: Put inside the if above, so we don't check it twice when it's not empty.

File: addons/sys_intercom/vic3/data/Readme.md [hunk 4]
```diff
@@ -0,0 +1 @@
+VIC3 FFCS models donated by RanTa to the ACRE Team. Any use outside ACRE project is explicitly forbidden without authorization.
```
Comment (jonpas) [inline addons/sys_intercom/vic3/data/Readme.md]: Same as GM stuff, we should modify license to support that and, like it is done in ACE3.

Separate component like ACE3's `apl`, with a different license that forbids any non-ACRE2 usage.

File: addons/sys_intercom/vic3/fnc_vic3ffcsConfig.sqf [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+/*
+ * Author: ACRE2Team
+ * Configures the VIC-3 FFCS station.
+ *
+ * Arguments:
+ * 0: Intercom Hash <ARRAY>
+ * 1: Intercom configuration <ARRAY>
+ *  0: Intercom volume <NUMBER>
+ *  1: Monitoring rack <NUMBER>
+ *  2: Actively used rack <NUBMER>
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsConfig.sqf]:
```suggestion
 *  2: Actively used rack <NUMBER>
```

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnIntercomKnobPress.sqf [hunk 6]
```diff
@@ -0,0 +1,51 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Handles turning the intercom knob.
+ *
+ * Arguments:
+ * 0: Unused
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnIntercomKnobPress.sqf]:
```suggestion
 * 0: What it actually is <TYPE> (unused)
```

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnIntercomKnobPress.sqf [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ * 1: Key direction. O: left, 1: right <NUMBER>
+ *
+ * Return Value:
+ * RETURN VALUE <TYPE>
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnIntercomKnobPress.sqf]:
```suggestion
 * None
```

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnMonitorKnobPress.sqf [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+/*
+ * Author: ACRE2Team
+ * Handles turning the monitor knob.
+ *
+ * Arguments:
+ * 0: Unused
+ * 1: Key direction. O: left, 1: right <NUMBER>
+ *
+ * Return Value:
+ * RETURN VALUE <TYPE>
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnMonitorKnobPress.sqf]:
```suggestion
 * None
```

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnMonitorKnobPress.sqf [hunk 9]
```diff
@@ -0,0 +1,96 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Handles turning the monitor knob.
+ *
+ * Arguments:
+ * 0: Unused
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnMonitorKnobPress.sqf]: Same as above.

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnVolumeKnobPress.sqf [hunk 10]
```diff
@@ -0,0 +1,36 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Handles turning the volume knob.
+ *
+ * Arguments:
+ * 0: Unused
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnVolumeKnobPress.sqf]: Same as above.

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnWorkKnobPress.sqf [hunk 11]
```diff
@@ -0,0 +1,76 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Handles turning the work knob.
+ *
+ * Arguments:
+ * 0: Unused
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnWorkKnobPress.sqf]: Same as above.

File: addons/sys_intercom/vic3/fnc_vic3ffcsOnWorkKnobPress.sqf [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+ * 1: Key direction. O: left, 1: right <NUMBER>
+ *
+ * Return Value:
+ * RETURN VALUE <TYPE>
```
Comment (jonpas) [inline addons/sys_intercom/vic3/fnc_vic3ffcsOnWorkKnobPress.sqf]:
```suggestion
 * None
```

File: addons/sys_intercom/vic3/vic3_ffcs_intercomDialog.hpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+#define CONTROL_SetRelativePos(xpos,ypos) x = H_OFFSET + (xpos * 0.001); y = H_OFFSET + (ypos * 0.001);
+#define CONTROL_SetDimensions(width, height) w = width * 0.001; h = height * 0.001;
+
+#define BEGIN_CONTROL(name, parent, idval) class name: parent { idc = idval;
+#define END_CONTROL };
+
+#define NEW_SCALE 1.4
+#define SCALE (NEW_SCALE*1)
+
+class VIC3FFCS_IntercomDialog {
```
Comment (jonpas) [inline addons/sys_intercom/vic3/vic3_ffcs_intercomDialog.hpp:12]:
```suggestion
class ACRE_VIC3FFCS_IntercomDialog {
```
And wherever it is used.

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp [hunk 14]
```diff
@@ -0,0 +1,62 @@
+class VIC3FFCS_RscBackground {
```
Comment (jonpas) [inline addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:1]:
```suggestion
class ACRE_VIC3FFCS_RscBackground {
```

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    type = CT_STATIC;
+    idc = -1;
+    style = ST_BACKGROUND + ST_KEEP_ASPECT_RATIO;
+    colorBackground[] = {0, 0, 0, .75};
+    colorText[] = {1, 1, 1, 1};
+    font = FontM;
+    sizeEx = 0.04;
+};
+
+class VIC3FFCS_RscButton {
```
Comment (jonpas) [inline addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:11]:
```suggestion
class ACRE_VIC3FFCS_RscButton {
```

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    soundPush[] = {"", 0.2, 1};
+    soundClick[] = {"", 0.2, 1};
+    soundEscape[] = {"", 0.2, 1};
+
+    default = false;
+    text = "";
+    action = "";
+};
+
+class VIC3FFCS_RscPicture: VIC3FFCS_RscButton {
```
Comment (jonpas) [inline addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:57]:
```suggestion
class ACRE_VIC3FFCS_RscPicture: ACRE_VIC3FFCS_RscButton {
```

**Review by jonpas — CHANGES_REQUESTED** [review]

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp (hunk 14 above)
Comment (TheMagnetar) [reply addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:1]: None of the radio dialogs have what you suggested. For coherence I would keep it like it is now.

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp (hunk 15 above)
Comment (TheMagnetar) [reply addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:11]: Same as above

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp (hunk 16 above)
Comment (TheMagnetar) [reply addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:57]: Same as above

File: addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp (hunk 14 above)
Comment (jonpas) [reply addons/sys_intercom/vic3/vic3ffcs_dialogClasses.hpp:1]: Let's hope no one else decides to ever use the same name? :P

**Review by jonpas — APPROVED** [review]

**Pull request merged**